From 2303f24ee7fb119d9ab3dad1fb6eff3cf911b1ee Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sat, 5 Oct 2024 06:34:55 -0700
Subject: [PATCH v2] configure.ac: Omit frame pointers and align stack on clang.

Align the stack to 4 bytes for PE targets on clang, same as -mpreferred-stack-boundary=2 for gcc.
The previous version of this patch incorrectly used -mstack-alignment=2, but clang interprets this
as "n" bytes, while gcc uses "2^n" bytes.

https://linux.kernel.narkive.com/y7hAplfS/patch-x86-build-specify-stack-alignment-for-clang

---
diff --git a/configure.ac b/configure.ac
index 11111111111..11111111111 100644
--- a/configure.ac
+++ b/configure.ac
@@ -963,6 +963,7 @@ This is an error since --enable-archs=$wine_arch was requested.])])
     WINE_TRY_PE_CFLAGS([-Wabsolute-value])
     WINE_TRY_PE_CFLAGS([-Wenum-enum-conversion],[:],WINE_TRY_PE_CFLAGS([-Wenum-conversion]))
     WINE_TRY_PE_CFLAGS([-ffunction-sections])
+    WINE_TRY_PE_CFLAGS([-fomit-frame-pointer])
 
     dnl clang had broken -fms-hotpatch support before version 18 (https://github.com/llvm/llvm-project/pull/77245)
     WINE_TRY_PE_CFLAGS([-fms-hotpatch -DMIN_CLANG_VERSION=18],
@@ -973,8 +974,8 @@ This is an error since --enable-archs=$wine_arch was requested.])])
     WINE_TRY_PE_CFLAGS([-flarge-source-files -Wmisleading-indentation],[AS_VAR_APPEND(${wine_arch}_EXTRACFLAGS,[" -Wno-misleading-indentation"])])
 
     case $wine_arch in
-      i386) WINE_TRY_PE_CFLAGS([-fno-omit-frame-pointer])
-            WINE_TRY_PE_CFLAGS([-mpreferred-stack-boundary=2])
+      i386) WINE_TRY_PE_CFLAGS([-mstack-alignment=4])
+            WINE_TRY_PE_CFLAGS([-mpreferred-stack-boundary=2])
             WINE_TRY_PE_CFLAGS([-Wl,--disable-stdcall-fixup],
                                [AS_VAR_APPEND([${wine_arch}_LDFLAGS],[" -Wl,--disable-stdcall-fixup"])]) ;;
       x86_64) WINE_TRY_PE_CFLAGS([-Wformat-overflow])
@@ -1968,7 +1969,6 @@ then
 
   case $HOST_ARCH in
     dnl gcc-4.6+ omits frame pointers by default, breaking some copy protections
-    i386) WINE_TRY_CFLAGS([-fno-omit-frame-pointer],[MSVCRTFLAGS="$MSVCRTFLAGS -fno-omit-frame-pointer"]) ;;
     x86_64)
       case $host_os in
         dnl Mingw uses Windows 64-bit types, not Unix ones
@@ -1987,6 +1987,7 @@ int a(int b, ...) { __builtin_ms_va_list list; __builtin_ms_va_start(list,b); }]
       esac ;;
     arm)
       WINE_TRY_CFLAGS([-Wincompatible-function-pointer-types],[EXTRACFLAGS="$EXTRACFLAGS -Wno-error=incompatible-function-pointer-types"]) ;;
+    *) WINE_TRY_CFLAGS([-fomit-frame-pointer],[MSVCRTFLAGS="$MSVCRTFLAGS -fomit-frame-pointer"]) ;;
   esac
 
   CFLAGS=$saved_CFLAGS
-- 
0.0.0

